core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/controller/promisehandlers/SkewDetectionHandler.scala

```diff
@@ -126,56 +249,287 @@ object SkewDetectionHandler {
+    * `skewedAndHelperWorkersList`.
+    */
+  private def implementFirstPhasePartitioning[T](
+      prevWorkerLayer: WorkerLayer,
+      skewedWorker: ActorVirtualIdentity,
+      helperWorker: ActorVirtualIdentity
+  ): Future[Seq[Boolean]] = {
+
+    val futures = new ArrayBuffer[Future[Boolean]]()
+    prevWorkerLayer.workers.keys.foreach(id => {
+      futures.append(
+        send(
+          SharePartition(
+            skewedWorker,
+            helperWorker,
+            Constants.reshapeFirstPhaseSharingNumerator,
+            Constants.reshapeFirstPhaseSharingDenominator
+          ),
+          id
+        )
+      )
+    })
+
+    Future.collect(futures)
+  }
+
+  private def implementSecondPhasePartitioning[T](
+      prevWorkerLayer: WorkerLayer,
+      skewedWorker: ActorVirtualIdentity,
+      helperWorker: ActorVirtualIdentity
+  ): Future[Seq[Boolean]] = {
+    val futures = new ArrayBuffer[Future[Boolean]]()
+    prevWorkerLayer.workers.keys.foreach(id => {
+      if (
+        workloadSamples.contains(id) && workloadSamples(id)
+          .contains(skewedWorker) && workloadSamples(id)
+          .contains(helperWorker)
+      ) {
+        // Second phase requires that the samples for both skewed and helper workers
+        // are recorded at the previous worker `id`. This will be used to partition the
+        // incoming data for the skewed worker.
+        var skewedLoad = AmberUtils.predictedWorkload(workloadSamples(id)(skewedWorker))
+        var helperLoad = AmberUtils.predictedWorkload(workloadSamples(id)(helperWorker))
+        var redirectNumerator = ((skewedLoad - helperLoad) / 2).toLong
+        workloadSamples(id)(skewedWorker) = new ArrayBuffer[Long]()
+        workloadSamples(id)(helperWorker) = new ArrayBuffer[Long]()
+        if (skewedLoad == 0 || helperLoad > skewedLoad) {
+          helperLoad = 0
+          skewedLoad = 1
+          redirectNumerator = 0
+        }
+        futures.append(
+          send(
+            SharePartition(
+              skewedWorker,
+              helperWorker,
+              redirectNumerator,
+              skewedLoad.toLong
+            ),
+            id
+          )
+        )
+
+      }
+    })
+
+    Future.collect(futures)
+  }
+
+  private def implementPauseMitigation[T](
+      prevWorkerLayer: WorkerLayer,
+      skewedWorker: ActorVirtualIdentity,
+      helperWorker: ActorVirtualIdentity
+  ): Future[Seq[Boolean]] = {
+    val futuresArr = new ArrayBuffer[Future[Boolean]]()
+    prevWorkerLayer.workers.keys.foreach(id => {
+      futuresArr.append(send(PauseSkewMitigation(skewedWorker, helperWorker), id))
+    })
+    Future.collect(futuresArr)
+  }
+
   registerHandler((msg: ControllerInitiateSkewDetection, sender) => {
-    if (!previousCallFinished) {
+    if (
+      !previousSkewDetectionCallFinished || !firstPhaseRequestsFinished || !secondPhaseRequestsFinished || !pauseMitigationRequestsFinished
```

Comment:
Hi sorry the logic is too long and I would not be very suitable to comment about this without a deep understanding. Some comment thoughts:
1. To simplify logic and review, shall we consider make those sub steps into separate unit functions/methods, thus it is more modularized and it's easier to maintain or optionally enable/disable a step.
2. I see all onSuccess handlers. Shall we also add onFailure handler to handle possible exceptions during each step, or in general handle any errors as a failure of  skew detection, and optionally revert to a stable state?

Reply:
Resolved as Shengquan is handling this PR